src/prototype.rs

```diff
@@ -210,6 +241,86 @@ impl<'a> FirstPass<'a> {
         ix
     }
 
+    /// When start_ix is at the beginning of an HTML block of type 1 to 5,
+    /// this will find the end of the block, adding the block itself to the
+    /// tree and also keeping track of the lines of HTML within the block.
+    ///
+    /// The html_end_tag is the tag that must be found on a line to end the block.
+    fn parse_html_block_type_1_to_5(&mut self, start_ix: usize, html_end_tag: &'static str) -> usize {
+        self.tree.append(Item {
+            start: start_ix,
+            end: 0, // set later
+            body: ItemBody::HtmlBlock(Some(html_end_tag)),
+        });
+        self.tree.push();
+
+        let mut ix = start_ix;
+        let end_ix;
+        loop {
+            let line_start_ix = ix;
+            ix += scan_nextline(&self.text[ix..]);
+            self.append_html_line(line_start_ix, ix);
+
+            let mut line_start = LineStart::new(&self.text[ix..]);
+            let n_containers = self.scan_containers(&mut line_start);
+            if n_containers < self.tree.spine.len() {
+                end_ix = ix;
+                break;
+            }
+
+            if (&self.text[line_start_ix..ix]).contains(html_end_tag) {
+                end_ix = ix;
+                break;
+            }
+
+            let next_line_ix = ix + line_start.bytes_scanned();
+            if next_line_ix == self.text.len() {
+                end_ix = next_line_ix;
+                break;
+            }
+            ix = next_line_ix;
+        }
+        &self.pop(end_ix);
+        ix
+    }
+
+    /// When start_ix is at the beginning of an HTML block of type 6 or 7,
+    /// this will consume lines until there is a blank line and keep track of
+    /// the HTML within the block.
+    fn parse_html_block_type_6_or_7(&mut self, start_ix: usize) -> usize {
+        self.tree.append(Item {
+            start: start_ix,
+            end: 0, // set later
+            body: ItemBody::HtmlBlock(None)
+        });
+        self.tree.push();
+
+        let mut ix = start_ix;
+        let end_ix;
+        loop {
+            let line_start_ix = ix;
+            ix += scan_nextline(&self.text[ix..]);
+            self.append_html_line(line_start_ix, ix);
+
+            let mut line_start = LineStart::new(&self.text[ix..]);
+            let n_containers = self.scan_containers(&mut line_start);
+            if n_containers < self.tree.spine.len() || line_start.is_at_eol()
+            {
+                end_ix = ix;
+                break;
+            }
+
+            let next_line_ix = ix + line_start.bytes_scanned();
+            if next_line_ix == self.text.len() || scan_eol(&self.text[next_line_ix..]).1 {
```

Comment:
`scan_eol` isn't exactly the detection of a blank line, it will fail to catch a line consisting of whitespace.

Reply:
:-$

I suppose `scan_blank_line()` would be more appropriate.